
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Appendix &#8212; miRNAture 1.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="License" href="license.html" />
    <link rel="prev" title="Tutorial" href="tutorial.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="appendix">
<h1>Appendix<a class="headerlink" href="#appendix" title="Permalink to this heading">¶</a></h1>
<section id="pre-processing-mirbase-data">
<h2>Pre-processing miRBase data:<a class="headerlink" href="#pre-processing-mirbase-data" title="Permalink to this heading">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">MIRfix</span></code> pipeline <a href="#id1"><span class="problematic" id="id2">:cite:p:`Yazbeck:19a`</span></a> provides the general core of
functions to curate <em>bona fide</em> metazoan microRNA annotations. To make use of
this curation process, it is fundamental to organize the input data in a
specific format, as referenced in more detail in <a href="#id3"><span class="problematic" id="id4">:cite:`Yazbeck:19a`</span></a>. In
summary, it is required:</p>
<blockquote>
<div><ul class="simple">
<li><p>A set of precursor sequences with their associated mature sequences.</p></li>
<li><p>Genome sequences from which miRNAs were annotated.</p></li>
<li><p>A relation file that describes the relation between precursors and their
annotated matures.</p></li>
</ul>
</div></blockquote>
<p>Additional parameters are required, but they did not depend from external
information/databases.</p>
<p>On <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> the source of the curation data has been obtained from a re-evaluation
of the annotations deposited on <code class="docutils literal notranslate"><span class="pre">miRBase</span></code> v.22.1 <a href="#id5"><span class="problematic" id="id6">:cite:`Kozomara:19`</span></a>. In this
version, <code class="docutils literal notranslate"><span class="pre">miRBase</span></code> accounted for a set of X <em>canonical</em> and <em>non-canonical</em> miRNA families,
from which Y are constituted by metazoan sequences. Internally, <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code>
performs an evaluation of the <em>canonical</em> model, that relies on the correct
positioning cleavages performed by Drosha and Dicer at the precursor maturation steps.
Computationally, this is translated on a correct position of the mature sequence,
accurate delimitation of precursor, and a phylogenetic support, addressed by the
construction of family <em>mature-anchored</em> structural alignments. As previously
reported for the <code class="docutils literal notranslate"><span class="pre">Rfam</span></code> miRNA families <a href="#id7"><span class="problematic" id="id8">:cite:`VelandiaDiss:2022`</span></a>, an iterative
assessment involves a selection of sequences, consistent criteria to evaluate
the miRNAs and their mature products, and generation of probabilistic models
derived from anchored-alignments to search additional candidates that would
incorporate defined curation criteria. This criteria was inherited to perform an
evaluation of the <code class="docutils literal notranslate"><span class="pre">miRBase</span></code> metazoan families and generate the corrected
dataset that <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> uses to evaluate new candidates and their maturation
entities.</p>
<p>As a toy example, the family miR-17 (MIPF0000001) was selected to demonstrate
the assessment steps performed over all pre-calculated dataset used by <cite>miRNAture</cite>.
As reported in <code class="docutils literal notranslate"><span class="pre">miRBase</span></code> this family is composed by 239 miRNA precursors derived from
39 vertebrate species. Through the filtering approach the following subsetting steps are
considered:</p>
<ul class="simple">
<li><p>Remove non-metazoan sequences.</p></li>
<li><p>Filter duplicates (which share 100% identity) and select one representant
sequence.</p></li>
</ul>
<p>In this family, 117 duplicated sequences where recognized. For instance the
sequence bta-mir-18a (MI0004740) from <em>Bos taurus</em> has shown 21 orthologs, as follows:</p>
<figure class="align-center" id="id22">
<span id="table-dup"></span><a class="reference internal image-reference" href="_images/table_MI0004740.png"><img alt="Orthologs in other species" src="_images/table_MI0004740.png" style="width: 600px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 4 </span><span class="caption-text">Identified orthologs from bta-mir-18a on vertebrates.</span><a class="headerlink" href="#id22" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>And some corresponding alignments:</p>
<figure class="align-center" id="id23">
<span id="align-dup"></span><a class="reference internal image-reference" href="_images/alignments_MI0004740.png"><img alt="Orthologs in other species" src="_images/alignments_MI0004740.png" style="width: 600px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 5 </span><span class="caption-text">Alignments as evidence of 100% identity.</span><a class="headerlink" href="#id23" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>Remaining 122 families were subject of a structural assessment by <code class="docutils literal notranslate"><span class="pre">MIRfix</span></code>,
which filtered 4 sequences based on the incorrect miRNA folding in regard their
annotated mature sequences, and one sequence contained a bad positioned mature
sequence in the reported precursor, a successful extension of the precursor based
on the miR and miR* prediction, rescued the candidate.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Category</p></th>
<th class="head"><p>Accession numbers</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Bad position mature sequences</p></td>
<td><p>MI0004822</p></td>
</tr>
<tr class="row-odd"><td><p>Filtered sequences</p></td>
<td><p>MI0012797, MI0012947, MI0019542, and MI0013837</p></td>
</tr>
</tbody>
</table>
<p>At the end of the assessment 118 sequences passed all filters to be considered
into the curation dataset used on <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code>.</p>
<p>The same approach curated all metazoan miRNA families from <code class="docutils literal notranslate"><span class="pre">miRBase</span></code> (1415),
validating about 79% (1111) of the families and setting the curation dataset used on
<code class="docutils literal notranslate"><span class="pre">miRNAture</span></code>.</p>
</section>
<section id="construction-of-hidden-markov-and-covariance-models">
<h2>Construction of Hidden Markov and Covariance Models:<a class="headerlink" href="#construction-of-hidden-markov-and-covariance-models" title="Permalink to this heading">¶</a></h2>
<p>As described in <a href="#id9"><span class="problematic" id="id10">:cite:`Velandia:2021`</span></a>, a set of quality-filtering steps could be
used to construct family structural alignments and their corresponding
covariance models (CMs). In this case, to build new structural alignments from
<code class="docutils literal notranslate"><span class="pre">miRBase</span></code> sequences, we selected all sequences from metazoan
species and removed all of those from studied organisms. Given that curated
subset, a genetic algorithm was used to maximize the quality the final structural
alignment. To do so, filtering miRNA sequences was done in function of: Identity
percentage (<span class="math notranslate nohighlight">\(I\)</span>), phylogenetic distribution of sequences (<span class="math notranslate nohighlight">\(C\)</span>) and quality
(<span class="math notranslate nohighlight">\(Q\)</span>) <a class="footnote-reference brackets" href="#f3" id="id11" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>, where: <span class="math notranslate nohighlight">\(I = (60, 70, 80, 90, 100)\)</span>, <span class="math notranslate nohighlight">\(C =\)</span> (Metazoa, Vertebrata,
Mammalia, Primates) and <span class="math notranslate nohighlight">\(Q =\)</span> (normal, high). An individual <span class="math notranslate nohighlight">\(A_{n}\)</span> was
defined as a vector <span class="math notranslate nohighlight">\(\overrightarrow{A_{n}}= \begin{pmatrix} I \\ C \\ Q
\end{pmatrix}\)</span>, which return a structural alignment using <code class="docutils literal notranslate"><span class="pre">MIRfix</span></code>, using selected
sequences. The <em>fitness</em> function (<span class="math notranslate nohighlight">\(F\)</span>) to be maximized was defined through
empirical observation over features inferred from generated structural alignment,
as follows:</p>
<div class="math notranslate nohighlight">
\[F = (N_{seq} + (N_{spe} * (-F_{energy})) + (N_{parts} * 10))\]</div>
<p>Where <span class="math notranslate nohighlight">\(N_{seq}\)</span> is the final number of sequences, <span class="math notranslate nohighlight">\(N_{spe}\)</span> is the number of species,
<span class="math notranslate nohighlight">\(F_{energy}\)</span> corresponds to folding energy calculated using
<code class="docutils literal notranslate"><span class="pre">RNAalifold</span></code> <a href="#id12"><span class="problematic" id="id13">:cite:`Lorenz2011`</span></a> and <span class="math notranslate nohighlight">\(N_{parts}\)</span> accounts the number of
additional (<span class="math notranslate nohighlight">\(&gt; 1\)</span>) stem-loops on the reported consensus structure. The initial
population was <span class="math notranslate nohighlight">\(A_{p}=40\)</span>, used operators were: <em>Selection</em> = Tournament,
<span class="math notranslate nohighlight">\(n=39\)</span>; <em>Crossover</em> = Single point, probability=0.7; <em>Mutation</em> =
Displacement mutation, probability=0.1. The implementation were performed in
<code class="docutils literal notranslate"><span class="pre">Python</span></code> v3.7.9 using <code class="docutils literal notranslate"><span class="pre">deap</span></code> package <a href="#id14"><span class="problematic" id="id15">:cite:`Fortin:2012`</span></a>.</p>
<p>Finally, hidden Markov (HMMs) and covariance (CMs) models were build as
described in <a href="#id16"><span class="problematic" id="id17">:cite:`Velandia:2021`</span></a> using <code class="docutils literal notranslate"><span class="pre">RNAalifold</span></code> <a href="#id18"><span class="problematic" id="id19">:cite:`Lorenz2011`</span></a>
and <code class="docutils literal notranslate"><span class="pre">Infernal</span></code> package v.1.1.2 :cite:<a href="#id20"><span class="problematic" id="id21">``</span></a>.</p>
<p class="rubric">Footnotes</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="f3" role="note">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id11">1</a><span class="fn-bracket">]</span></span>
<p>Confidence of the annotation assigned by <code class="docutils literal notranslate"><span class="pre">miRBase</span></code>, see <cite>https://www.mirbase.org/blog/2014/03/high-confidence-micrornas/</cite></p>
</aside>
</aside>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">miRNAture</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick_start.html">Quick Start</a></li>
<li class="toctree-l1"><a class="reference internal" href="strategy.html">Strategy</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Appendix</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Need Help</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="tutorial.html" title="previous chapter">Tutorial</a></li>
      <li>Next: <a href="license.html" title="next chapter">License</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021, miRNAture.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.1.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/appendix.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>