
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Introduction &#8212; miRNAture 1.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Installation" href="installation.html" />
    <link rel="prev" title="miRNAture" href="index.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="introduction">
<h1>Introduction<a class="headerlink" href="#introduction" title="Permalink to this heading">Â¶</a></h1>
<p>MicroRNAs (miRNAs) have been characterized as important regulators
in almost all animals and plants, as well as in unicellular eukaryotes <a href="#id1"><span class="problematic" id="id2">:cite:`Moran:17`</span></a>.
Since their discovery in 1993 <a href="#id3"><span class="problematic" id="id4">:cite:`Lee1993`</span></a> and subsequently their recognition as
a biological entity later from 2000 <a href="#id5"><span class="problematic" id="id6">:cite:`Reinhart2000`</span></a>, microRNAs were identified
as key regulators on the temporal control of heterochronic genes on the nematode
<em>Caenorhabditis elegans</em> as well as in another metazoan species, too <a href="#id7"><span class="problematic" id="id8">:cite:`Pasquinelli:00`</span></a>.
This recognition was complemented by a complete characterization of their
typical features as: a stem-loop structure, well-conservation over multiple metazoan clades
and typical expression patterns as isolated locus or co-expressed as polycistronic miRNA
transcripts <a href="#id9"><span class="problematic" id="id10">:cite:`Lagos-Quintana:2001,Lau:2001,Lee:2001,Reinhart2000`</span></a>.</p>
<p>Currently, metazoan miRNAs have been recognized as a conserved group of short ncRNAs,
typically ~ 22 nt, with important roles as post-transcriptional regulators of the gene expression
affecting a sizeable number of mRNAs and expressed in all developmental
process and diseases <a href="#id11"><span class="problematic" id="id12">:cite:`Bartel:18`</span></a>. Their canonical biogenesis starts in the form of primary
precursors (pri-miRNAs) transcribed from long non-coding RNAs or protein-coding transcripts,
mostly from introns <a href="#id13"><span class="problematic" id="id14">:cite:`Zeidler:2020`</span></a>. Later, derived from hairpin-like precursors excised in
the nucleus (pre-miRNAs), their acting form is subsequently further processed as
miRNA/miRNA* duplexes on the cytoplasm and incorporated into the RISC complex.
Target specificity is achieved by complementarity between the miR and mRNA sequence.
(see more details in <a href="#id15"><span class="problematic" id="id16">:cite:`Bartel:18`</span></a>).</p>
<p>Current classification of annotated miRNAs into families are available in miRBase <a class="footnote-reference brackets" href="#f1" id="id17" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a> and
mirGeneDB <a class="footnote-reference brackets" href="#f2" id="id18" role="doc-noteref"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></a> databases. As an example, the human genome reported 1984 miRNA precursors in
the miRBase v.22.1 <a href="#id19"><span class="problematic" id="id20">:cite:`Kozomara:19`</span></a> and the corresponding mature products were estimated ~2300
<a href="#id21"><span class="problematic" id="id22">:cite:`Alles:19`</span></a>. Focusing on the <em>confidently canonical</em> miRNAs reported in <a href="#id23"><span class="problematic" id="id24">:cite:`Fromm:15`</span></a>,
the number of miRNAs is 519. Those differences are explained on the basis of multiple
<em>miRNA</em> detection methods as well as the intrinsic definitions to define a <em>canonical miRNA</em>.</p>
<p>Despite the small size of the precursors (80-100 nt), sequence comparison
methods are able to detected them, due a high level of sequence conservation
<a href="#id25"><span class="problematic" id="id26">:cite:`Price:11`</span></a>. In one hand, it is important to point that the use of blast-based homology
searches alone tend to produce false positives that require extensive curation, which relies
on properties obtained from miRNAs <a href="#id27"><span class="problematic" id="id28">:cite:`Tarver:18`</span></a>. On the other hand, the
inclusion of the consensus structure complements the homology search methods,
for example using covariance models (CMs) <a href="#id29"><span class="problematic" id="id30">:cite:`Eddy:94, Gardner:2009`</span></a>. The accuracy and
sensitivity of CMs depends critically on the sequence alignment and the annotated consensus
used to build the model. Those observations call for an integrated workflow to perform
homology search and to evaluate their results consistently.</p>
<p>In this computational approach, focused on the <em>canonical</em> miRNAs processed by Drosha and Dicer,
we improve on ideas from MIRfix <a href="#id31"><span class="problematic" id="id32">:cite:`Yazbeck:19a`</span></a> and integrate it with homology search.
<code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> is used to identify and annotate homologs of metazoan microRNAs in a homology-based
setting.</p>
<p class="rubric">Footnotes</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="f1" role="note">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id17">1</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="https://www.mirbase.org/">https://www.mirbase.org/</a></p>
</aside>
<aside class="footnote brackets" id="f2" role="note">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id18">2</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="https://www.mirgenedb.org/">https://www.mirgenedb.org/</a></p>
</aside>
</aside>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">miRNAture</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="strategy.html">Strategy</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="appendix.html">Appendix</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Need Help</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">miRNAture</a></li>
      <li>Next: <a href="installation.html" title="next chapter">Installation</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021, miRNAture.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.1.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/introduction.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>