
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Quick Start &#8212; miRNAture 1.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Strategy" href="strategy.html" />
    <link rel="prev" title="Installation" href="installation.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="quick-start">
<h1>Quick Start<a class="headerlink" href="#quick-start" title="Permalink to this heading">¶</a></h1>
<section id="input-files">
<h2>Input files<a class="headerlink" href="#input-files" title="Permalink to this heading">¶</a></h2>
<p>The most important input file is a DNA sequence. This could be a multifasta
sequence that belongs from a common specie (i.e. complete genome or group of
particular sequences). At the same time, previous to execute <strong>miRNAture</strong> a
you have to download a pre-calculated dataset (as indicated on <a class="reference internal" href="#precalculated-data"><span class="std std-ref">Pre-calculated datasets</span></a> Section)
that contains default data as CMs, HMMs, and required files to perform mature prediction.
Once located in your computer, the path might be indicated with the flag <code class="docutils literal notranslate"><span class="pre">-dataF</span></code>.</p>
<section id="activate-the-mirnature-environment">
<h3>Activate the <code class="docutils literal notranslate"><span class="pre">mirnature</span></code> environment<a class="headerlink" href="#activate-the-mirnature-environment" title="Permalink to this heading">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">activate</span> <span class="n">mirnature</span>
</pre></div>
</div>
</section>
<section id="run-mirnature">
<h3>Run <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code><a class="headerlink" href="#run-mirnature" title="Permalink to this heading">¶</a></h3>
<p>A complete mode should be run as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">./</span><span class="n">miRNAture</span> <span class="o">-</span><span class="n">stage</span> <span class="n">complete</span> <span class="o">-</span><span class="n">dataF</span> <span class="o">&lt;</span><span class="n">Precalculated_folder</span><span class="o">&gt;</span> \
            <span class="o">-</span><span class="n">speG</span> <span class="o">&lt;</span><span class="n">Target</span> <span class="n">Genome</span><span class="o">&gt;</span> <span class="o">-</span><span class="n">speN</span> <span class="o">&lt;</span><span class="n">Specie_name</span><span class="o">&gt;</span> \
            <span class="o">-</span><span class="n">speT</span> <span class="o">&lt;</span><span class="n">Tag_specie</span><span class="o">&gt;</span> <span class="o">-</span><span class="n">w</span> <span class="o">&lt;</span><span class="n">Output_dir</span><span class="o">&gt;</span> \
            <span class="o">-</span><span class="n">m</span> <span class="o">&lt;</span><span class="n">Mode</span><span class="o">&gt;</span> <span class="p">(</span><span class="o">-</span><span class="nb">str</span> <span class="o">&lt;</span><span class="n">Blast_strategy</span><span class="o">&gt;</span><span class="p">)</span> \
            <span class="o">-</span><span class="n">blastq</span> <span class="o">&lt;</span><span class="n">Blast_queries_folder</span><span class="o">&gt;</span>
</pre></div>
</div>
<p>But it is always recommended to look up specific parameters. Do not use the
default parameters for all experiments.</p>
</section>
<section id="output-files">
<h3>Output files<a class="headerlink" href="#output-files" title="Permalink to this heading">¶</a></h3>
<p>Final predicted miRNAs will be written on the <code class="docutils literal notranslate"><span class="pre">&lt;Output_dir&gt;</span></code> indicated with the <code class="docutils literal notranslate"><span class="pre">-w</span></code> flag.
The final candidates are described on the folder <code class="docutils literal notranslate"><span class="pre">Final_miRNA_evaluation/</span></code> as
follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Final_miRNA_evaluation/
├── Fasta/
├── MFE/
├── miRNA_annotation_&lt;Tag_specie&gt;_accepted_conf.bed
├── miRNA_annotation_&lt;Tag_specie&gt;_accepted_conf.gff3
├── miRNAture_summary_&lt;Tag_specie&gt;.txt
└── Tables/
</pre></div>
</div>
<p>Inside this folder, <strong>miRNAture</strong> will create 3 folders containing their
correspondent results: sequences in <code class="docutils literal notranslate"><span class="pre">fasta</span></code> format (<code class="docutils literal notranslate"><span class="pre">Fasta/</span></code>), minimum free energy and
lengths from described sequences (<code class="docutils literal notranslate"><span class="pre">MFE/</span></code>) and the supporting information ordered in tables
for each annotated candidate (<code class="docutils literal notranslate"><span class="pre">Tables/</span></code>). Additionally, associated genomic positions
for the miRNA candidates are reported in <code class="docutils literal notranslate"><span class="pre">BED</span></code> and <code class="docutils literal notranslate"><span class="pre">GFF3</span></code> formats and a summary file,
<code class="docutils literal notranslate"><span class="pre">miRNAture_summary_&lt;Tag_specie&gt;.txt</span></code>, that describes overall descriptive statistics from found miRNA
families.</p>
</section>
<section id="pre-calculated-datasets">
<span id="precalculated-data"></span><h3>Pre-calculated datasets<a class="headerlink" href="#pre-calculated-datasets" title="Permalink to this heading">¶</a></h3>
<p>Pre-calculated data composed by miRNA CMs, HMMs and required input files to perform mature annotation has
to be downloaded before run the full <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> pipeline. Available datasets are listed below:</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>NEW</strong>: Curated metazoan families from miRBase v.22.1, available to the structural validation stage.</p></li>
<li><p>Required data to re-annotate human miRNAs: include CMs and HMMs build from miRBase without human sequences.
Stored in Zenodo <a class="reference external" href="https://zenodo.org/record/4531376#.YCQS8EMo_ys">here</a>.</p></li>
</ul>
</div></blockquote>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">miRNAture</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Quick Start</a></li>
<li class="toctree-l1"><a class="reference internal" href="strategy.html">Strategy</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="appendix.html">Appendix</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Need Help</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="installation.html" title="previous chapter">Installation</a></li>
      <li>Next: <a href="strategy.html" title="next chapter">Strategy</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021, miRNAture.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.1.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/quick_start.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>