<!DOCTYPE html>
<html >
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
      <title>Searching miRNAs</title>
    
      <link rel="stylesheet" href="_static/pygments.css">
      <link rel="stylesheet" href="_static/theme.css">
      <link rel="stylesheet" href="_static/romnnn_sphinx_press_theme.css">
      
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>

      <!-- sphinx script_files -->
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>

      
      <script src="_static/theme-vendors.js"></script>
      <script src="_static/theme.js" defer></script>
    
  <link rel="index" title="Index" href="genindex.html" />
  <link rel="search" title="Search" href="search.html" /> 
  </head>

  <body>
    <div id="app" class="theme-container" :class="pageClasses"><navbar @toggle-sidebar="toggleSidebar">
  <div class="home-link-container">
      <router-link to="index.html" class="home-link">
      
        <span class="site-name">miRNAture</span>
    </router-link>
  </div>

  <div class="links">
    <navlinks class="can-hide">



    </navlinks>
  </div>
</navbar>

      
      <div class="sidebar-mask" @click="toggleSidebar(false)">
      </div>
        <sidebar @toggle-sidebar="toggleSidebar">
          
          <navlinks>
            



            
          </navlinks><div id="searchbox" class="searchbox" role="search">
  <div class="caption"><span class="caption-text">Search</span>
    <div class="searchformwrapper">
      <form autocomplete="off" class="search" action="search.html" method="get">
        <input type="text" name="q" autocomplete="off" />
        <input type="submit" value="Go" />
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div><div class="sidebar-links" role="navigation" aria-label="main navigation">
  
    <div class="sidebar-group">
      <p class="caption">
        <span class="caption-text"><a href="index.html#mirnature">Contents:</a></span>
      </p>
      <ul class="">
        
          <li class="toctree-l1 "><a href="introduction.html" class="reference internal ">Introduction</a>

            
          </li>

        
          <li class="toctree-l1 "><a href="installation.html" class="reference internal ">Installation</a>

            
          </li>

        
          <li class="toctree-l1 "><a href="strategy.html" class="reference internal ">Strategy</a>

            
          </li>

        
          <li class="toctree-l1 "><a href="tutorial.html" class="reference internal ">Tutorial</a>

            
          </li>

        
          <li class="toctree-l1 "><a href="license.html" class="reference internal ">License</a>

            
          </li>

        
          <li class="toctree-l1 "><a href="help.html" class="reference internal ">Need Help</a>

            
          </li>

        
      </ul>
    </div>
  
</div>
        </sidebar>

      <page>
          <div class="body-header" role="navigation" aria-label="navigation">
  
  <ul class="breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
    
    <li>Searching miRNAs</li>
  </ul>
  

  <ul class="page-nav">
</ul>
  
</div>
<hr>
          <div class="content" role="main">
            
  <div class="section" id="searching-mirnas">
<h1>Searching miRNAs<a class="headerlink" href="#searching-mirnas" title="Permalink to this headline">¶</a></h1>
<p>The most important step will be performed! Based on the last configurations on
the script <code class="docutils literal notranslate"><span class="pre">run_miRNAture.sh</span></code>, <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> will be executed on the designed
coelacanth sequences. The idea is to perform independently each of the stages,
<em>homology-searches</em> and <em>detection of mature</em>, for demonstrative
purposes. In case that you require to run the complete pipeline, just adjust the
parameter <strong>-stage</strong> to the <em>complete</em> option.</p>
<div class="section" id="homology-search">
<h2>Homology search<a class="headerlink" href="#homology-search" title="Permalink to this headline">¶</a></h2>
<p>As mentioned, we are going to execute the <em>homology-search</em> stage. To activate
this stage in <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> please verify the flag value to be <strong>-stage homology</strong>.
In brief, our <em>target</em> coelacanth sequences would be annotated using a set of
miRNA <em>queries</em> that belong from the following chordate species, (V: vertebrata, T:
tunicata and C:cephalochordata) and one echinoderm (E):</p>
<ul class="simple">
<li><p><em>Anolis carolinensis</em> (V)</p></li>
<li><p><em>Branchiostoma belcheri</em> (C)</p></li>
<li><p><em>Branchiostoma floridae</em> (C)</p></li>
<li><p><em>Ciona robusta</em> (T)</p></li>
<li><p><em>Ciona savignyi</em> (T)</p></li>
<li><p><em>Danio rerio</em> (V)</p></li>
<li><p><em>Eptatretus burgeri</em> (V)</p></li>
<li><p><em>Petromyzon marinus</em> (V)</p></li>
<li><p><em>Strongylocentrotus purpuratus</em> (E)</p></li>
<li><p><em>Xenopus laevis</em> (V)</p></li>
<li><p><em>Xenopus tropicalis</em> (V)</p></li>
</ul>
<p>We are going to test all the capability of <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code>, using at the same time
all the available modes: <code class="docutils literal notranslate"><span class="pre">Blast</span></code>, <code class="docutils literal notranslate"><span class="pre">HMM</span></code>, <code class="docutils literal notranslate"><span class="pre">Infernal</span></code> and the final
concatenation with <code class="docutils literal notranslate"><span class="pre">Final</span></code>. Specifically for the pairwise-comparisons with
<code class="docutils literal notranslate"><span class="pre">Blast</span></code> mode, we are going to use only 3 strategies: 1,9,10 and the final
concatenation and comparison with <code class="docutils literal notranslate"><span class="pre">ALL</span></code>, but feel free to choose more or less
strategies.</p>
<p>Then, just let it run typing:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$./run_miRNAture.sh homology
</pre></div>
</div>
<p>A long descriptive output will be printed on the screen. Keep an eye on the <code class="docutils literal notranslate"><span class="pre">Results/</span></code>
folder where the action is taking place.</p>
<blockquote>
<div><div class="admonition note">
<p class="admonition-title">Note</p>
<p>Keep in mind that <code class="docutils literal notranslate"><span class="pre">run_miRNAture.sh</span></code> was created as an example to run <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> but it does not mean that is the only way to do that. Change it according your requirements.</p>
</div>
</div></blockquote>
</div>
<div class="section" id="homology-search-results">
<h2>Homology search results<a class="headerlink" href="#homology-search-results" title="Permalink to this headline">¶</a></h2>
<p>To refer directly to the results, type:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd ../Results/
$ tree -L 1
Results/
├── Blast
├── Final_Candidates
├── HMMs
├── Infernal
├── miRNAture_log_190609072022.log
├── miRNAture_log_22505108072023.log
├── miRNAture_log_23590008072025.log
├── miRNAture_log_4209072026.log
└── mirnature_runLatch.sh
</pre></div>
</div>
<p>If everything goes well, you could see 4 log files <code class="docutils literal notranslate"><span class="pre">miRNA_log_*.log</span></code>, a
script generated automatically to run the search strategy on <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code>
(<code class="docutils literal notranslate"><span class="pre">mirnature_runLatch.sh</span></code>) and the folders with homology comparisons: <code class="docutils literal notranslate"><span class="pre">Blast/</span></code>,
<code class="docutils literal notranslate"><span class="pre">HMMs</span></code> and direct structure comparison: <code class="docutils literal notranslate"><span class="pre">Infernal</span></code> and the <code class="docutils literal notranslate"><span class="pre">Final_Candidates</span></code>
with the final set of homology predicted miRNAs. Next, go directly to the
<code class="docutils literal notranslate"><span class="pre">Final_Candidates</span></code> folder:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd Final_Candidates/
$ tree -L 1
├── all_RFAM_Latch_Final.ncRNAs_homology.txt
├── all_RFAM_Latch_Final.truetable
├── all_RFAM_Latch_Final.truetable.discarded.table
├── all_RFAM_Latch_Final.truetable.joined.table
├── all_RFAM_Latch_Final.truetable.joined.table.db
├── all_RFAM_Latch_Final.truetable.temp
└── Fasta
</pre></div>
</div>
<p>Where the most important file is <code class="docutils literal notranslate"><span class="pre">all_RFAM_Latch_Final.ncRNAs_homology.txt</span></code>,
which reported all the merged candidates to miRNAs on the subject contigs
from coelacanth. The results are summarised on the following table:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 27%" />
<col style="width: 21%" />
<col style="width: 23%" />
<col style="width: 29%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Contig</p></th>
<th class="head"><p>Length (Mb)</p></th>
<th class="head"><p>Numb. miRNAs</p></th>
<th class="head"><p>miRNAture Pred.</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>JH126571.1</p></td>
<td><p>5.98145</p></td>
<td><p>5</p></td>
<td><p>122</p></td>
</tr>
<tr class="row-odd"><td><p>JH126620.1</p></td>
<td><p>3.03251</p></td>
<td><p>8</p></td>
<td><p>106</p></td>
</tr>
<tr class="row-even"><td><p>AFYH01291077.1</p></td>
<td><p>0.00106</p></td>
<td><p>1</p></td>
<td><p>0</p></td>
</tr>
</tbody>
</table>
<p>The final results could be discriminated by the annotation method (Blast, HMM or
Infernal):</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 31%" />
<col style="width: 11%" />
<col style="width: 7%" />
<col style="width: 18%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Contig</p></th>
<th class="head"><p>Blast</p></th>
<th class="head"><p>HMM</p></th>
<th class="head"><p>Infernal</p></th>
<th class="head"><p>miRNAture Pred.</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>JH126571.1</p></td>
<td><p>22</p></td>
<td><p>7</p></td>
<td><p>93</p></td>
<td><p>122</p></td>
</tr>
<tr class="row-odd"><td><p>JH126620.1</p></td>
<td><p>35</p></td>
<td><p>6</p></td>
<td><p>65</p></td>
<td><p>106</p></td>
</tr>
</tbody>
</table>
<p>and even, this set of computational annotations could be visualized on a broad
genome context, generating for example a <a class="reference external" href="https://www.ensembl.org/info/website/upload/bed.html">BED</a> file and uploading it at the Coelacanth Ensembl Genome Browser, using some Linux commands as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$awk &#39;{print $1&quot;\t&quot;$6&quot;\t&quot;$7&quot;\t&quot;$8&quot;\t&quot;$2&quot;\t&quot;$3}&#39;
all_RFAM_Latch_Final.ncRNAs_homology.txt &gt; predicted_miRNAture.bed
</pre></div>
</div>
<p>Next, just upload the track on the corresponding Genome Ensembl hub (as explained in more detail <a class="reference external" href="https://www.ensembl.org/info/website/upload/index.html">here</a>) as a Custom Track.</p>
<p>Certainly, after uploading this miRNAs coordinates you would visualize this results:</p>
<div class="align-center figure">
<a class="reference internal image-reference" href="_images/general_view.png"><img alt="Genome Browser Coel" src="_images/general_view.png" style="width: 600px; height: 99px;" /></a>
</div>
<p>This image corresponds to the region <code class="docutils literal notranslate"><span class="pre">JH126620.1:788915-822338</span></code>, which
according to the Ensembl annotation, exists 4 miRNA genes
of the families: mir-130 (ENSLACG00000021481), mir-454 (ENSLACG00000019256),
and two of mir-130 (ENSLACG00000020131, ENSLACG00000019975)</p>
<p>Here, <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> detected the same families, with overlapping regions on the
previously reported miRNAs on <em>L. chalumnae</em>.</p>
<div class="align-center figure">
<a class="reference internal image-reference" href="_images/firstTwo.png"><img alt="Genome Browser Coel" src="_images/firstTwo.png" style="width: 600px; height: 99px;" /></a>
</div>
<p>And this is the second cluster, with two families, the overlapping is the same
but in one miRNA the family prediction and the strand differ:</p>
<div class="align-center figure">
<a class="reference internal image-reference" href="_images/secondTwo.png"><img alt="Genome Browser Coel" src="_images/secondTwo.png" style="width: 600px; height: 99px;" /></a>
</div>
<p>For that reason, those candidates required a complementary evaluation of their current detection and correct positioning of the <em>mature</em> miRNA sequences.
As a final result, you could check that all the reported miRNAs on the contigs
JH126571.1 and JH126620.1 were identified. The reported miRNA on AFYH01291077.1,
was predicted as a miRNA using RFAM, but currently there is no information about
the family or mature products. <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> detected this candidate on the
direct Infernal searches, but it did not show an acceptable homology (for
mir-105 family) and folding values (Bitscore: 13.4 and E-value 7.8), see file
<code class="docutils literal notranslate"><span class="pre">Results/miRNA_predictionInfernal/Latch/RF00670_Latch.tab</span></code>.</p>
<p>On the other side, <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> detected new candidates that currently are not
reported on the genome annotation.</p>
</div>
<div class="section" id="validation-of-mirna-candidates">
<h2>Validation of miRNA candidates<a class="headerlink" href="#validation-of-mirna-candidates" title="Permalink to this headline">¶</a></h2>
<p>An additional output was generated on the <code class="docutils literal notranslate"><span class="pre">Final_results/</span></code> folder and contained
all the resulted fasta sequences from the last 228 hairpin candidates, organized
by their Rfam family. Based on those regions, validated by sequence and
structure homology, the idea is to evaluate the annotation of candidate <em>mature</em>
regions that are contained in this hairpin-loop and validate their annotation
with an additional layer, supported by the structural alignments of sequences
selected from other organisms.</p>
<p>To do that, please execute again the script:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$./run_miRNAture.sh validate
</pre></div>
</div>
<p>which essentially have the same input parameters, except for the <code class="docutils literal notranslate"><span class="pre">-stage</span> <span class="pre">validation</span></code> flag
that was changed to tell <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> that the second stage have to be
activated.</p>
<p>In this step, each detected miRNA candidate were grouped by their Rfam miRNA family.
Based on this reference, previously calculated data from the family is retrieved. This
input data, required to perform the correction of the <em>mature</em> sequences using the
<code class="docutils literal notranslate"><span class="pre">MIRfix</span></code> program, was inferred as a product of this study <a class="footnote-reference brackets" href="#f1" id="id1">1</a> and
comprises this set of files:</p>
<blockquote>
<div><ul class="simple">
<li><p>The set of Rfam hairpin sequences.</p></li>
<li><p>The mature sequences annotated for each Rfam hairpin sequence.</p></li>
<li><p>The genomes/contigs/sequences that contained the Rfam sequences.</p></li>
<li><p>A mapping file, which explicity declares the relation between hairpin and
their mature sequences.</p></li>
</ul>
</div></blockquote>
<p>For more details refer to the <code class="docutils literal notranslate"><span class="pre">MIRfix</span></code> <a class="reference external" href="https://github.com/Bierinformatik/MIRfix">repository</a> .
Automatically, <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> structure all your data and generate the required
input files to perform the <em>mature</em> annotation.</p>
<dl class="footnote brackets">
<dt class="label" id="f1"><span class="brackets"><a class="fn-backref" href="#id1">1</a></span></dt>
<dd><p>From the <em>seed</em> sequences from <code class="docutils literal notranslate"><span class="pre">Rfam</span></code> v.12.2 and additionally the sequences from <code class="docutils literal notranslate"><span class="pre">Rfam</span></code> and <code class="docutils literal notranslate"><span class="pre">miRBase</span></code> that reported <em>mature</em> sequences.</p>
</dd>
</dl>
</div>
</div>


          </div>
          <div class="page-nav">
            <div class="inner"><ul class="page-nav">
</ul><div class="footer" role="contentinfo">
      &#169; Copyright 2021, miRNAture.
    <br>
    Created using <a href="http://sphinx-doc.org/">Sphinx</a> 3.5.1 with <a href="https://github.com/romnnn/sphinx_press_theme">Press Theme</a>.
</div>
            </div>
          </div>
      </page>
    </div>
    
    
  </body>
</html>