
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Strategy &#8212; miRNAture 1.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorial" href="tutorial.html" />
    <link rel="prev" title="Installation" href="installation.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="strategy">
<h1>Strategy<a class="headerlink" href="#strategy" title="Permalink to this heading">¶</a></h1>
<p>Current approaches to computational miRNA detection relies on homology relationships
or detection of hairpin-loop candidates with lower folding energy. A complete set of
tools to automatize this task have been assembled on <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code>. This current
approach combines two different sequence-homology modes, using <code class="docutils literal notranslate"><span class="pre">blast</span></code> or <code class="docutils literal notranslate"><span class="pre">HMMer</span></code>, and
a secondary structure validation step, performed by the <code class="docutils literal notranslate"><span class="pre">INFERNAL</span></code> package. Merging and
consolidating task from multiple search strategies is done automatically by <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code>,
throwing at the the end of the <em>Homology searches</em> stage, a list of regions that
reported highest scores based on selected homology searches. Those candidates
passed designed filters (see in more detail in <a class="reference internal" href="#filters"><span class="std std-numref">Fig. 1</span></a>), to be considered as homologs by the
applied computational searches.</p>
<p>Further structural and microRNA-specific evaluations are covered on the
<em>Mature evaluation</em> step, which makes use of an updated version of the original <code class="docutils literal notranslate"><span class="pre">MIRfix</span></code>
pipeline <a href="#id1"><span class="problematic" id="id2">:cite:`Yazbeck:19a`</span></a>. At this step, <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> evaluates the identity at
family level of the homology candidates found in the previous step. Based on
that, makes use of the reported precursor, mature and genomic information contained on the
<code class="docutils literal notranslate"><span class="pre">miRBase</span></code> database. Specially for this step, this curation step is prepared
and reported with each release, allowing the user to perform the best mature positioning
and assignment on their predicted precursor sequences. Please refer to the
<a class="reference internal" href="appendix.html#appendix"><span class="std std-ref">Appendix</span></a> section to know more details about the curation process of
the <code class="docutils literal notranslate"><span class="pre">miRBase</span></code> database and the generation of associated data. At this point,
for each of those precursors, <code class="docutils literal notranslate"><span class="pre">MIRfix</span></code> will try to:</p>
<blockquote>
<div><ul class="simple">
<li><p>Assign the best-fitting mature sequence from those reported for the
discovered miRNA family.</p></li>
<li><p>Predict the position of the miR* and correct the precursor sequence, based on
the assigned mature.</p></li>
<li><p>Evaluate on a multiple structural alignment, the fit of the new annotated precursor in
regard existing annotated miRNAs classified in the same family.</p></li>
</ul>
</div></blockquote>
<p>Those results are feeded onto additional evaluation steps, that would assign a
confidence to each precursor with its associated mature(s), namely: High, Medium
or No confidence.</p>
<figure class="align-center" id="id3">
<span id="filters"></span><a class="reference internal image-reference" href="_images/methodmirnature.png"><img alt="miRNAture filters" src="_images/methodmirnature.png" style="width: 600px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 1 </span><span class="caption-text">Designed homology/strucrure filters in <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code>. Specific programs used for each mode in parenthesis. Ann.: Annotation, SS: Secondary structure. CSS: Consensus secondary structure. ge: gathering cutoff from Rfam family. nBit = Bitscore/ge. ted: tree edit distance between default miRNA and modified multiple stockholm alignments. MFE: Minimum free energy. HSPs: high scoring pairs.</span><a class="headerlink" href="#id3" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>In the <em>complete</em> mode, <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> will report the following output files:</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">GFF3</span></code> and <code class="docutils literal notranslate"><span class="pre">BED</span></code> files of the precursors with their mature sequences.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Fasta</span></code> sequences from miRNA precurspors.</p></li>
<li><p>A summary table describing features of found miRNAs, such as: their <em>loci</em> number, family
classification and their confidence.</p></li>
</ul>
</div></blockquote>
<p>Current workflow is depicted in <a class="reference internal" href="#workflow"><span class="std std-numref">Fig. 2</span></a>:</p>
<figure class="align-center" id="id4">
<span id="workflow"></span><a class="reference internal image-reference" href="_images/mirnature3.png"><img alt="``miRNAture`` workflow" src="_images/mirnature3.png" style="width: 600px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 2 </span><span class="caption-text">General <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> workflow.</span><a class="headerlink" href="#id4" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">miRNAture</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Strategy</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="appendix.html">Appendix</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Need Help</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="installation.html" title="previous chapter">Installation</a></li>
      <li>Next: <a href="tutorial.html" title="next chapter">Tutorial</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021, miRNAture.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.1.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/strategy.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>