
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Tutorial &#8212; miRNAture 1.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Appendix" href="appendix.html" />
    <link rel="prev" title="Strategy" href="strategy.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="tutorial">
<h1>Tutorial<a class="headerlink" href="#tutorial" title="Permalink to this heading">¶</a></h1>
<p>Through this step-by-step tutorial you could make use of key options from
<code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> to annotate the <em>bona fide</em> miRNA complement on selected contigs
from the coelacanth (<em>Latimeria chalumnae</em>) genome. All the required files to
execute this tutorial are included in the <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> source files in the <cite>miRNAture/Tutorial</cite>
folder.</p>
<section id="annotating-let-7-on-coelacanth">
<h2>Annotating let-7 on coelacanth<a class="headerlink" href="#annotating-let-7-on-coelacanth" title="Permalink to this heading">¶</a></h2>
<figure class="align-center" id="id3">
<a class="reference internal image-reference" href="_images/coel.jpg"><img alt="Coelacanth image" src="_images/coel.jpg" style="width: 320px; height: 116px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 3 </span><span class="caption-text"><em>Latimeria chalumnae</em>. Source: <a class="reference external" href="https://upload.wikimedia.org/wikipedia/commons/f/fa/Latimeria_Chalumnae_-_Coelacanth_-_NHMW.jpg">Alberto Fernandez Fernandez / CC BY-SA</a></span><a class="headerlink" href="#id3" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>Based on the miRNA annotation retrieved from <code class="docutils literal notranslate"><span class="pre">Ensembl</span></code> release 100, the
coelacanth genome featured 8 let-7 locus distributed along 6 contigs. For purposes of this
tutorial, 3 contigs were selected with variable number of miRNA/let-7 annotations, as follows:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Contig</p></th>
<th class="head"><p>Length (Mb)</p></th>
<th class="head"><p>Numb. miRNAs</p></th>
<th class="head"><p>Let-7 loci</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>JH126571.1</p></td>
<td><p>5.981</p></td>
<td><p>5</p></td>
<td><p>1</p></td>
</tr>
<tr class="row-odd"><td><p>JH129429.1</p></td>
<td><p>0.248</p></td>
<td><p>3</p></td>
<td><p>3</p></td>
</tr>
<tr class="row-even"><td><p>AFYH01291077.1</p></td>
<td><p>0.001</p></td>
<td><p>1</p></td>
<td><p>0</p></td>
</tr>
</tbody>
</table>
<p>The main goal is the identification of the let-7 loci on the referred
contigs. To do so, the <cite>Tutorial</cite> folder contained all the input files in <cite>Data/</cite>,
required wrapper to <cite>miRNAture</cite> on <cite>Code/</cite> and a <cite>Results/</cite> folder where
all prediction will be stored. As you can imagine, homology comparisons are prone
to create both, a high number of input and output files. <cite>miRNAture</cite> avoids manual
curation to the detected hits, life is too short to perform all of those steps by hand!</p>
<section id="folder-structure">
<h3>Folder structure<a class="headerlink" href="#folder-structure" title="Permalink to this heading">¶</a></h3>
<p>The folder tree on miRNAture looks like:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ tree -L 1 miRNAture/
miRNAture/
├── Build.PL
├── Changes
├── ignore.txt
├── index.md
├── lib/
├── LICENSE
├── MANIFEST
├── META.json
├── META.yml
├── mirnature_logo.png
├── miRNAture-Manual/
├── miRNAture.yml
├── README
├── README.md
├── README.rst
├── script/
├── t/
├── Tutorial/
└── xt/
</pre></div>
</div>
<p>Our target folder is located in <code class="docutils literal notranslate"><span class="pre">Tutorial/</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd Tutorial/
$ tree -L2 .
Tutorial/
├── Code
│   ├── list_miRNAs_to_search.txt
│   ├── Precalculated-Data-tutorial
│   ├── tutorial_test_selected_models.sh
│   └── User_Test_Data
├── Data
│   ├── latimeria_chalumnae_genome.fa
│   └── QueriesToTest
└── Results
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">Tutorial</span></code> folder is composed by the subfolders: <code class="docutils literal notranslate"><span class="pre">Code/</span></code>, where all
the necessary scripts to run <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> are located. <code class="docutils literal notranslate"><span class="pre">Data/</span></code> keeps the described
contigs from coelacanth in a multi-fasta file:<code class="docutils literal notranslate"><span class="pre">latimeria_chalumnae_genome.fa</span></code>. In the same
folder, in <code class="docutils literal notranslate"><span class="pre">QueriesToTest/</span></code> let-7 annotations from 11 metazoans <a class="footnote-reference brackets" href="#species" id="id1" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a> were provided
as queries.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Together with the query files, the file <cite>queries_description.txt</cite> is required to
control which dataset of sequences that will be used by the <cite>blastn</cite> comparisons. Three
columns are needed to be recognized:</p>
<p>&lt;Name_fasta_file.fa&gt; miRNA &lt;Origin_of_sequence&gt;</p>
<p>The first one corresponds to the file name, the second one have to be miRNA, the third
one is the name of the source specie in the format: <cite>Genera specie</cite>. If you do not know
the source, a valid name would be: Unknown specie. If ommited, <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> will create
automatically this file using all fasta files in this folder with an Unknown origin.</p>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">Results/</span></code> folder will conserve all the output files generated by <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code>.</p>
</section>
<section id="input-files">
<h3>Input files<a class="headerlink" href="#input-files" title="Permalink to this heading">¶</a></h3>
<p>To run <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> just go directly to <code class="docutils literal notranslate"><span class="pre">Code/</span></code> folder:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd Code/
$ tree -L 1 .
.
├── list_miRNAs_to_search.txt
├── Precalculated-Data-tutorial
├── tutorial_test_selected_models.sh
└── User_Test_Data
</pre></div>
</div>
<p>In this path, the <code class="docutils literal notranslate"><span class="pre">tutorial_test_selected_models.sh</span></code> file is <code class="docutils literal notranslate"><span class="pre">bash</span></code> script that will organize
all our code to run <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code>. This way is preferred in terms of reproducibility means of your
computational experiments. This code will give you a general idea to run <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code>, let’s explain
this in detail:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/bin/bash</span>

<span class="nv">current</span><span class="o">=</span><span class="k">$(</span> <span class="nb">pwd</span> <span class="k">)</span>
<span class="nv">specie_tag</span><span class="o">=</span><span class="s2">&quot;Lach&quot;</span>
<span class="nv">specie_genome</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$current</span><span class="s2">/../Data/latimeria_chalumnae_genome.fa&quot;</span>
<span class="nv">specie_name</span><span class="o">=</span><span class="s2">&quot;Latimeria_chalumnae&quot;</span>

<span class="nv">workdir</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$current</span><span class="s2">/../Results&quot;</span>
mkdir -p <span class="nv">$workdir</span>
<span class="nv">mode</span><span class="o">=</span><span class="s2">&quot;Blast,HMM,Infernal,Other_CM,Final&quot;</span>
<span class="nv">strategy</span><span class="o">=</span><span class="s2">&quot;5,6,ALL&quot;</span>
<span class="nv">blastQueriesFolder</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$current</span><span class="s2">/../Data/QueriesToTest&quot;</span>
<span class="nv">user_models</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$current</span><span class="s2">/User_Test_Data&quot;</span>
<span class="nv">data_precalculated_folder</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$current</span><span class="s2">/Precalculated-Data-tutorial&quot;</span>

<span class="c1">### Step by step: homology-&gt;validation-&gt;evaluation-&gt;summarise</span>
<span class="c1"># Run only homology-searches</span>
<span class="c1">#miRNAture -stage homology -sublist $current/list_miRNAs_to_search.txt \</span>
<span class="c1">#    -dataF $data_precalculated_folder -speG $specie_genome -speN $specie_name \</span>
<span class="c1">#    -speT $specie_tag -w $workdir -m $mode -pe 0 -str $strategy \</span>
<span class="c1">#    -blastq $blastQueriesFolder -rep relax,150,100 -usrM $user_models</span>
<span class="c1"># Run detection matures</span>
<span class="c1">#miRNAture -stage validation -dataF $data_precalculated_folder -speG $specie_genome \</span>
<span class="c1">#    -speN $specie_name -speT $specie_tag -w $workdir -m $mode -pe 0 -usrM $user_models</span>
<span class="c1"># Run the complete analysis</span>
<span class="c1">#miRNAture -stage evaluation -dataF $data_precalculated_folder -speG $specie_genome \</span>
<span class="c1">#    -speN $specie_name -speT $specie_tag -w $workdir -m $mode -pe 0</span>
<span class="c1"># Create summarise report</span>
<span class="c1">#miRNAture -stage summarise -dataF $data_precalculated_folder -speG $specie_genome \</span>
<span class="c1">#    -speN $specie_name -speT $specie_tag -w $workdir -m $mode -pe 0</span>

<span class="c1"># Run miRNAture complete</span>
miRNAture -stage <span class="nb">complete</span> -sublist <span class="nv">$current</span>/list_miRNAs_to_search.txt <span class="se">\</span>
    -dataF <span class="nv">$data_precalculated_folder</span> -speG <span class="nv">$specie_genome</span> -speN <span class="nv">$specie_name</span> <span class="se">\</span>
    -speT <span class="nv">$specie_tag</span> -w <span class="nv">$workdir</span> -m <span class="nv">$mode</span> -pe <span class="m">0</span> -str <span class="nv">$strategy</span> <span class="se">\</span>
    -blastq <span class="nv">$blastQueriesFolder</span> -rep relax,150,100 -usrM <span class="nv">$user_models</span>
</pre></div>
</div>
<p>Activate the <code class="docutils literal notranslate"><span class="pre">conda</span></code> environment called miRNAture. The installation and activation of this environment is required previously to run <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code>.  All the dependences are described on the file <code class="docutils literal notranslate"><span class="pre">miRNAture.yml</span></code>, located on the <code class="docutils literal notranslate"><span class="pre">miRNAture/Code/</span></code> folder.</p>
<p>The last script shows two steps that are required to run <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code>:</p>
<p>Declare the name of input and output locations. This will help to assign <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> flags and easily reproduce the experiment. In this case, we used the following options (flags indicated in parenthesis):</p>
<blockquote>
<div><ul class="simple">
<li><p>Processing stage (<code class="docutils literal notranslate"><span class="pre">-stage</span></code>): Running stage on <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code>. In this case was selected <code class="docutils literal notranslate"><span class="pre">complete</span></code> to run all the stages. To run step by step, this flag accepts: <code class="docutils literal notranslate"><span class="pre">homology</span></code>, <code class="docutils literal notranslate"><span class="pre">validation</span></code>, <code class="docutils literal notranslate"><span class="pre">evaluation</span></code> and <code class="docutils literal notranslate"><span class="pre">summarise</span></code>. You should run all of them in this order to obtain the same final results as <code class="docutils literal notranslate"><span class="pre">complete</span></code> option.</p></li>
<li><p>Subset of miRNA models to run (<code class="docutils literal notranslate"><span class="pre">-sublist</span></code>): Subset of miRNA families references to be searched on the target sequence. See <code class="docutils literal notranslate"><span class="pre">list_miRNAs_to_search.txt</span></code> file as an example. If not provided all miRNA RFAM models will be searched.</p></li>
<li><p>Pre-calculated data location (<code class="docutils literal notranslate"><span class="pre">-dataF</span></code>): Location of pre-calculated data required by <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code>. It included hidden markov, covariance models and curated input files to annotate mature sequences <a class="footnote-reference brackets" href="#importantnote" id="id2" role="doc-noteref"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></a>.</p></li>
<li><p>Specie genome (<code class="docutils literal notranslate"><span class="pre">-speG</span></code>): Current target sequence.</p></li>
<li><p>Specie name (<code class="docutils literal notranslate"><span class="pre">-speN</span></code>): Scientific name of the specie which belongs the subject sequence(s).</p></li>
<li><p>Specie tag (<code class="docutils literal notranslate"><span class="pre">-speT</span></code>): Tag of the specie name, suggested one takes the first two letters from the Genera joined with the first two from the specie (i.e <cite>Homo sapiens</cite> = Hosa).</p></li>
<li><p>Working directory (<code class="docutils literal notranslate"><span class="pre">-w</span></code>): Output directory, final path of <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> results.</p></li>
<li><p>Running mode (<code class="docutils literal notranslate"><span class="pre">-m</span></code>): Select at least one, or any combination of the miRNA search strategies between: <code class="docutils literal notranslate"><span class="pre">Blast</span></code>, <code class="docutils literal notranslate"><span class="pre">HMM</span></code>, <code class="docutils literal notranslate"><span class="pre">Infernal</span></code> and <code class="docutils literal notranslate"><span class="pre">Other_CM</span></code>. At the same time, to merge the complete results from those homology search modes, write at the end <code class="docutils literal notranslate"><span class="pre">Final</span></code>.</p></li>
<li><p>Parallel jobs using SLURM (<code class="docutils literal notranslate"><span class="pre">-pe</span></code>): Activate (1) or not (0).</p></li>
<li><p>Blast strategies (<code class="docutils literal notranslate"><span class="pre">-str</span></code>): Write the numbers of desired <code class="docutils literal notranslate"><span class="pre">blastn</span></code> strategies.  Possible strategies are: <code class="docutils literal notranslate"><span class="pre">1,2,3,4,5,6</span></code>. To merge all results put at the end <code class="docutils literal notranslate"><span class="pre">ALL</span></code>.</p></li>
<li><p>Path of <code class="docutils literal notranslate"><span class="pre">blastn</span></code> queries (<code class="docutils literal notranslate"><span class="pre">-blastq</span></code>): Declare the path of annotated query sequences of miRNAs. In this case is enough to indicate the folder name.</p></li>
<li><p>Homology repetition detection (<code class="docutils literal notranslate"><span class="pre">-rep</span></code>): Setup number of maximum loci number that will be evaluated by the mature annotation stage. By default, miRNAture will detect miRNA families that report high number of loci (&gt; 200 loci). Then, it will select the top 100 candidates in terms of alignment scores, as candidates for the validation stage (<code class="docutils literal notranslate"><span class="pre">default,200,100</span></code>). Modify this values using <code class="docutils literal notranslate"><span class="pre">relax,Number_Loci,Candidates_to_evaluate</span></code>.</p></li>
<li><p>User hidden markov/covariance models (<code class="docutils literal notranslate"><span class="pre">-usrM</span></code>): Directory with additional hidden Markov (HMMs) or covariance models (CMs) provided by the user to be searched on the target sequence.</p></li>
</ul>
</div></blockquote>
<p>Then, run <code class="docutils literal notranslate"><span class="pre">miRNAture</span></code> through this script:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ./tutorial_test_selected_models.sh
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The list of complete flags can be found typing: <code class="docutils literal notranslate"><span class="pre">miRNAture</span> <span class="pre">-h</span></code> or <code class="docutils literal notranslate"><span class="pre">miRNAture</span> <span class="pre">-man</span></code>.</p>
</div>
<p class="rubric">Footnotes</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="species" role="note">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">1</a><span class="fn-bracket">]</span></span>
<p><em>Anolis carolinensis</em>, <em>Branchiostoma belcheri</em>, <em>Branchiostoma floridae</em>, <em>Ciona robusta</em>, <em>Ciona savignyi</em>, <em>Danio rerio</em>, <em>Eptatretus burgeri</em>, <em>Petromyzon marinus</em>, <em>Strongylocentrotus purpuratus</em>, <em>Xenopus laevis</em> and <em>Xenopus tropicalis</em>.</p>
</aside>
<aside class="footnote brackets" id="importantnote" role="note">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id2">2</a><span class="fn-bracket">]</span></span>
<p>Pre-calculated data should be downloaded from <a class="reference external" href="https://zenodo.org/record/4531376#.YDqO4-bTVTZ">https://zenodo.org/record/4531376#.YDqO4-bTVTZ</a></p>
</aside>
</aside>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">miRNAture</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="strategy.html">Strategy</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="appendix.html">Appendix</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Need Help</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="strategy.html" title="previous chapter">Strategy</a></li>
      <li>Next: <a href="appendix.html" title="next chapter">Appendix</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021, miRNAture.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.1.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/tutorial.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>